pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -463,4 +493,145 @@ private long getWatermarkMs(long minStartTimeMs, long bucketMs, String mergeLeve
+    Map<String, Long> watermarksForTable = _mergeRollupWatermarks.remove(tableNameWithType);
+    if (watermarksForTable != null) {
+      for (String mergeLevel : watermarksForTable.keySet()) {
+        controllerMetrics.removeGauge(getMetricNameForTaskDelay(tableNameWithType, mergeLevel));
+      }
+    }
+  }
+
+  /**
+   * Reset the delay metrics for the given table name and merge level.
+   *
+   * @param tableNameWithType table name with type
+   * @param mergeLevel merge level
+   */
+  private void resetDelayMetrics(String tableNameWithType, String mergeLevel) {
+    ControllerMetrics controllerMetrics = _clusterInfoAccessor.getControllerMetrics();
+    if (controllerMetrics == null) {
+      return;
+    }
+
+    // Delete all the watermarks associated with the given the table name and the merge level.
+    Map<String, Long> watermarksForTable = _mergeRollupWatermarks.get(tableNameWithType);
+    if (watermarksForTable != null) {
+      if (watermarksForTable.remove(mergeLevel) != null) {
+        controllerMetrics.removeGauge(getMetricNameForTaskDelay(tableNameWithType, mergeLevel));
+      }
+    }
+  }
+
+  /**
+   * Clean up the metrics that no longer need to be emitted.
+   *
+   * We clean up the metrics for the following cases:
+   *   1. Table got deleted.
+   *   2. The current controller is no longer the leader for a table.
+   *   3. Merge task config got deleted.
+   *   4. Merge task config got modified and some merge levels got deleted.
+   *
+   * TODO: Current code will remove all metrics in case we invoke the ad-hoc task scheduling on a single table.
+   * We will file the follow-up PR to address this issue. We need to separate out APIs for ad-hoc scheduling and
+   * periodic scheduling. We will only enable metrics for periodic case.
+   *
+   * @param tableConfigs list of tables
+   */
+  private void cleanUpDelayMetrics(List<TableConfig> tableConfigs) {
+    Map<String, TableConfig> tableConfigMap = new HashMap<>();
+    for (TableConfig tableConfig : tableConfigs) {
+      tableConfigMap.put(tableConfig.getTableName(), tableConfig);
+    }
+
+    for (String tableNameWithType : new ArrayList<>(_mergeRollupWatermarks.keySet())) {
+      TableConfig currentTableConfig = tableConfigMap.get(tableNameWithType);
+      // Table does not exist in the cluster
+      if (currentTableConfig == null) {
+        resetDelayMetrics(tableNameWithType);
+        continue;
+      }
+
+      // The current controller is no longer leader for this table
+      if (!_clusterInfoAccessor.getLeaderControllerManager().isLeaderForTable(tableNameWithType)) {
+        resetDelayMetrics(tableNameWithType);
+        continue;
+      }
+
+      // Task config is removed
+      Map<String, String> taskConfigs = currentTableConfig.getTaskConfig().getConfigsForTaskType(getTaskType());
+      if (taskConfigs == null) {
+        resetDelayMetrics(tableNameWithType);
+        continue;
+      }
```

Comment:
This is not possible because only the table with task enabled can be passed into this class

Reply:
What if the task was enabled so the controller has gauge enabled and then the task config got removed later? I'm trying to cover that case. We call `cleanUpDelayMetrics()` at the end of the scheduling round.

Reply:
discussed offline.